### PR TITLE
Feature/auto resource cleanup

### PR DESCRIPTION
**Pull request by mihran113**

### CODE REVIEW

File: aim/ext/cleanup/__init__.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        ...
+
+    @staticmethod
+    def _cleanup():
+        """
+        Actually cleanup all the remaining objects in order of specified
+        priorities. This is called automatically by atexit.
+        It is not meant tobe called manually.
+        """
+        logger.debug("Cleaning up...  Found {len(finalizers)} finalizers")
```
Comment (mahnerak) [inline aim/ext/cleanup/__init__.py:72]: Missed `f'...`. Please fix the quotes.

File: aim/sdk/repo.py [hunk 2]
```diff
@@ -81,16 +108,18 @@ def __init__(self, path: str, *, read_only: bool = None, init: bool = False):
                 unmount_remote_repo(self.root_path, self._mount_root)
             raise RuntimeError(f'Cannot find repository \'{self.path}\'. Please init first.')
 
-        self.container_pool: Dict[ContainerConfig, Container] = WeakValueDictionary()
+        self.container_pool: WeakValueDictionary[ContainerConfig, Container] = WeakValueDictionary()
         self.persistent_pool: Dict[ContainerConfig, Container] = dict()
-        self.container_view_pool: Dict[ContainerConfig, Container] = WeakValueDictionary()
+        self.container_view_pool: WeakValueDictionary[ContainerConfig, Container] = WeakValueDictionary()
```
Comment (mahnerak) [inline aim/sdk/repo.py:113]: I remember having issues with genertic typings of `WeakValueDictionary`.
Please check these with Python3.6.

Comment (mahnerak) [reply aim/sdk/repo.py:113]: Yup, tests do not pass with Python 3.6. We can either use Dict or a more generic MutableMapping.

File: aim/ext/cleanup/__init__.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    def join(self):
+        while True:
+            try:
+                # Pressing Ctrl-C will raise KeyboardInterrupt and stop only
+                # the `Thread.join()` call. The target function of `RobustExec`
+                # will continue to run.
+                return super().join()
+            except KeyboardInterrupt:
+                logger.warning("Received Ctrl-C. Closing gracefully.")
```
Comment (mahnerak) [inline aim/ext/cleanup/__init__.py:29]: @gorarakelyan This fully blocks force-Ctrl-C-aborting.
We can use this implementation now, but we should think of a way of letting user to abort the running process immediately.

File: aim/sdk/repo.py (hunk 2 above)
Comment (mihran113) [reply aim/sdk/repo.py:113]: Tests do not pass, but somewhere else, not here. `WeakKeyDictionary` is the problematic one.

Comment (mahnerak) [reply aim/sdk/repo.py:113]: Yes, there was an issue when using `WeakKeyDictionary[]` typing with generic `T`. I believe we should avoid using `WeakValueDictionary` and `WeakKeyDictionary` as type annotations (including here).

Comment (mihran113) [reply aim/sdk/repo.py:113]: yeah, will change back to `Dict`

File: aim/ext/cleanup/__init__.py (hunk 1 above)
Comment (mihran113) [reply aim/ext/cleanup/__init__.py]: done ✅

File: aim/ext/cleanup/__init__.py (hunk 3 above)
Comment (gorarakelyan) [reply aim/ext/cleanup/__init__.py]: What about displaying prompt message asking whether user wants to terminate the process immediately?

**Review by alberttorosyan — APPROVED** [review]

**Pull request merged**